src/components/project_view/submit.vue

```diff
@@ -0,0 +1,472 @@
+  unexpected_files: string[] = [];
+  missing_files: string[] = [];
+  patterns_with_too_few_matches: ExpectedFilePatternMismatch[] = [];
+  patterns_with_too_many_matches: ExpectedFilePatternMismatch[] = [];
+
+  late_days_remaining: number | null = null;
+
+  current_user: User | null = null;
+
+  async created() {
+    this.current_user = await User.get_current();
+    let late_days = await User.get_num_late_days(this.course.pk, this.current_user.pk);
+    this.late_days_remaining = late_days.late_days_remaining;
+  }
+
+  get num_submissions_per_day() {
+    if (this.project.groups_combine_daily_submissions
+        && this.project.submission_limit_per_day !== null) {
+      return this.project.submission_limit_per_day * this.group.member_names.length;
+    }
+    return this.project.submission_limit_per_day;
+  }
+
+  get show_late_day_count(): boolean {
+    return this.course.num_late_days !== 0
+           || (this.late_days_remaining !== null && this.late_days_remaining !== 0);
+  }
+
+  get submission_might_use_late_day() {
+    if (this.late_days_remaining === null || this.late_days_remaining === 0) {
+      return false;
+    }
+    if (this.project.soft_closing_time === null) {
+      return false;
+    }
+    let deadline = moment(this.project.soft_closing_time);
+    if (this.group.extended_due_date !== null) {
+      deadline = moment(this.group.extended_due_date);
+    }
+
+    let late_days_used = this.group.late_days_used[this.current_user!.username];
+    if (late_days_used !== undefined) {
+      deadline = deadline.add(late_days_used, 'd');
+    }
+
+    return moment().isAfter(deadline);
+  }
+
+  get file_mismatches_present() {
+    return this.unexpected_files.length !== 0
+           || this.missing_files.length !== 0
+           || this.patterns_with_too_few_matches.length !== 0
+           || this.patterns_with_too_many_matches.length !== 0;
+  }
+
+  process_files(files: File[]) {
+    this.d_submit_progress = 0;
+    this.submitted_files = files;
+    this.submitted_filenames = files.map((file) => file.name);
+    this.validate_files(files);
+    (<Modal> this.$refs.confirm_submit_modal).open();
+  }
+
+  private validate_files(files: File[]) {
+    this.unexpected_files = [];
+    this.missing_files = [];
+    this.patterns_with_too_few_matches = [];
+    this.patterns_with_too_many_matches = [];
+
+    this.check_for_unexpected_files(files);
+    this.check_expected_file_counts(files);
+  }
+
+  private check_for_unexpected_files(files: File[]) {
+    let filtered: File[] = [];
+    for (let file of files) {
+      if (!this.file_matches_any_patterns(file)) {
+        this.unexpected_files.push(file.name);
+      }
+      else {
+        filtered.push(file);
+      }
+    }
+  }
+
+  private file_matches_any_patterns(file: File): boolean {
+    for (let expected_file of this.project.expected_student_files) {
+      if (matches(file.name, expected_file.pattern)) {
+        return true;
+      }
+    }
+
```

Comment:
There's an extra space here.

Reply:
Blank line or not before a return statement isn't something that necessarily needs to be standardized, but sure, I can take out this one.